### PR TITLE
Master sale computation commitment date jao

### PR DESCRIPTION
**Pull request by jao-odoo**
Task : https://www.odoo.com/web#id=34343&view_type=form&model=project.task&action=333&active_id=131&menu_id=4720

Pad : https://pad.odoo.com/p/r.0827fde0087f4d4e0efcf754a2941ab1

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/sale_order_dates/models/sale_order.py [hunk 1]
```diff
@@ -24,30 +24,28 @@ class SaleOrder(models.Model):
                                           "processed as soon as possible. In that case the expected "
                                           "date will be computed using the default method: based on "
                                           "the Product Lead Times and the Company's Security Delay.")
-    effective_date = fields.Date(compute='_compute_picking_ids', string='Effective Date', store=True,
-                                 help="Date on which the first Delivery Order was created.")
+    effective_date = fields.Date(compute='_compute_effactive_date', string='Effective Date', store=True,
```
Comment (mreficent) [inline addons/sale_order_dates/models/sale_order.py]: Why `effactive` instead of `effective`?

File: addons/sale_order_dates/tests/test_commitment_date.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+from odoo import tools
+from odoo import fields
+from datetime import timedelta
+
+
+class TestSaleCommitmentDates(common.TransactionCase):
+
+
+    def test_sale_order_commitment_date(self):
+        """ Test commitment date and effactive date """
```
Comment (mreficent) [inline addons/sale_order_dates/tests/test_commitment_date.py]: → effective

**Comment by rgo-odoo** [comment]
LGTM

File: addons/sale_order_dates/models/sale_order.py [hunk 3]
```diff
@@ -24,30 +24,28 @@ class SaleOrder(models.Model):
...
+                                 help="Date on which the first shipment was successfully delivered to customer.")
 
-    @api.depends('date_order', 'order_line.customer_lead')
+    @api.depends('order_line.customer_lead', 'confirmation_date', 'picking_policy')
     def _compute_commitment_date(self):
         """Compute the commitment date"""
         for order in self:
             dates_list = []
-            order_datetime = fields.Datetime.from_string(order.date_order)
+            confirm_date = fields.Datetime.from_string(order.confirmation_date or fields.Datetime.now())
```
Comment (tde-banana-odoo) [inline addons/sale_order_dates/models/sale_order.py]: Shouldn't it be based on create_date is no confirmation date ? Not sure, just asking.

File: addons/sale_order_dates/models/sale_order.py [hunk 4]
```diff
@@ -24,30 +24,28 @@ class SaleOrder(models.Model):
                                           "processed as soon as possible. In that case the expected "
                                           "date will be computed using the default method: based on "
                                           "the Product Lead Times and the Company's Security Delay.")
-    effective_date = fields.Date(compute='_compute_picking_ids', string='Effective Date', store=True,
-                                 help="Date on which the first Delivery Order was created.")
+    effective_date = fields.Date(compute='_compute_effective_date', string='Effective Date', store=True,
+                                 help="Date on which the first shipment was successfully delivered to customer.")
 
-    @api.depends('date_order', 'order_line.customer_lead')
+    @api.depends('order_line.customer_lead', 'confirmation_date', 'picking_policy')
```
Comment (tde-banana-odoo) [inline addons/sale_order_dates/models/sale_order.py]: I think we could add some triggers, like order_line.state, maybe create_date if we use it (not sure).

File: addons/sale_order_dates/models/sale_order.py [hunk 5]
```diff
@@ -24,30 +24,28 @@ class SaleOrder(models.Model):
...
-                dt = order_datetime + timedelta(days=line.customer_lead or 0.0)
+                dt = confirm_date + timedelta(days=line.customer_lead or 0.0)
                 dates_list.append(dt)
             if dates_list:
                 commit_date = min(dates_list) if order.picking_policy == 'direct' else max(dates_list)
                 order.commitment_date = fields.Datetime.to_string(commit_date)
 
-    def _compute_picking_ids(self):
-        super(SaleOrder, self)._compute_picking_ids()
+    @api.depends('order_line', 'order_line.qty_delivered')
```
Comment (tde-banana-odoo) [inline addons/sale_order_dates/models/sale_order.py]: Triggers seems a bit unrelated to the fields used in the method. Are you sure everything is present ?

File: addons/sale_order_dates/models/sale_order.py (hunk 4 above)
Comment (jir-odoo) [reply addons/sale_order_dates/models/sale_order.py]: We can set trigger order_line.state to set  commitment date false when state goes to cancel,
and we don't need create_date on trigger.

File: addons/sale_order_dates/models/sale_order.py (hunk 5 above)
Comment (jir-odoo) [reply addons/sale_order_dates/models/sale_order.py]: We can put here trigger "picking_ids.date_done" to call when shipment goes into done state and 
it will min(done_date) of done pickings.

File: addons/sale_order_dates/models/sale_order.py [hunk 6]
```diff
@@ -12,7 +12,7 @@ class SaleOrder(models.Model):
     commitment_date = fields.Datetime(compute='_compute_commitment_date', string='Commitment Date', store=True,
                                       help="Date by which the products are sure to be delivered. This is "
                                            "a date that you can promise to the customer, based on the "
-                                           "Product Lead Times.")
+                                           "Product Lead Times and the shipping policy.")
```
Comment (rim-odoo) [inline addons/sale_order_dates/models/sale_order.py]: remove `the`

File: addons/sale_order_dates/models/sale_order.py [hunk 7]
```diff
@@ -24,30 +24,28 @@ class SaleOrder(models.Model):
                                           "processed as soon as possible. In that case the expected "
                                           "date will be computed using the default method: based on "
                                           "the Product Lead Times and the Company's Security Delay.")
-    effective_date = fields.Date(compute='_compute_picking_ids', string='Effective Date', store=True,
-                                 help="Date on which the first Delivery Order was created.")
+    effective_date = fields.Date(compute='_compute_effective_date', string='Effective Date', store=True,
+                                 help="Date on which the first shipment was successfully delivered to customer.")
```
Comment (rim-odoo) [inline addons/sale_order_dates/models/sale_order.py]: s/was/is
